### PR TITLE
Add is_maximal_matching and is_matching functions

### LINKED ISSUE #255: Add `is_matching` and `is_maximal_matching` functions
### What is the expected enhancement?

In the tests added for #229 two utility functions, `is_matching` and `is_maximal_matching` were ported from networkx to work with retworkx (in python) and validate the result. We should include these functions to be native retworkx functions (in rust) and expose them through the retworkx api (and also update the tests to use that version instead).

### PR DESCRIPTION
**Pull request by mtreinish**
This commit adds new functions for checking a provided matching set is
valid, is_matching(), and that a provided matching set is valid and
maximal, is_maximal_set(). This pairs with #229 and can be used to
partially check the output from the max_weight_matching function added
there. Equivalent functions were implemented in #229 using Python for
the tests in #229 and those tests should be updated to use these
functions instead.

Fixes #255

### CODE REVIEW

**Review by itoko — CHANGES_REQUESTED** [review]
Thank you for a quick follow-up PR. I think we can simplify the implementation of `_inner_is_matching`.

File: src/lib.rs [hunk 1]
```diff
@@ -2659,6 +2659,126 @@ pub fn digraph_find_cycle(
...
+    let has_edge = |e: &(usize, usize)| -> bool {
+        graph
+            .graph
+            .contains_edge(NodeIndex::new(e.0), NodeIndex::new(e.1))
+    };
+
+    if !matching.iter().all(|e| has_edge(e)) {
+        return false;
+    }
+    let combinations = matching_combinations(matching);
```
[the remark this answers is not in the document]
Comment (mtreinish) [reply src/lib.rs:2709]: Oh, yeah that's a good idea. I was just looking at the networkx implementation that was using combinations and didn't think it through and just mirrored that here. But yyeah, just checking for duplicate endpoints by iterating over the list in the matching is should be sufficient and much less complicated.

[the remark this answers is not in the document]
Comment (mtreinish) [reply src/lib.rs]: Done in: https://github.com/Qiskit/retworkx/pull/256/commits/a81117b4975538dacc26c3ccb0e5397cb8bed374

File: src/lib.rs [hunk 2]
```diff
@@ -2659,6 +2659,126 @@ pub fn digraph_find_cycle(
...
+#[pyfunction]
+#[text_signature = "(graph, matching, /)"]
+pub fn is_matching(
+    graph: &graph::PyGraph,
+    matching: HashSet<(usize, usize)>,
+) -> bool {
+    _inner_is_matching(graph, &matching)
+}
+
+/// Check if a matching is a maximal matching for a graph
```
[the remark this answers is not in the document]
Comment (mtreinish) [reply src/lib.rs]: I added this in: https://github.com/Qiskit/retworkx/pull/256/commits/45c882a9d32612f5cd05ee1a7559c3d5784739fb let me know if you think I need to make it more clear.

**Review by itoko — APPROVED** [review]

File: src/lib.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (itoko) [reply src/lib.rs]: LGTM. Thanks!

**Pull request merged**